### ISSUE TITLE
Feedback from Ludwig

### ISSUE THREAD

**Issue by colah** [distillpub/post--memorization-in-rnns#6]
*Comments from Ludwig, posted by Chris since he doesn't have GitHub access*

## Consider small tweaks to Hero Diagram

A small suggestion, adding slightly more contrast to the text, rounded corners, turning down text contrast for names on the right...

![](https://storage.googleapis.com/ludwigschubert/images-misc/hero-colors.png)

## Consider weakening claims in "Connectivity in the Autocomplete Problem"

Consider hedging your language at the end of the "Connectivity in the Autocomplete Problem":

> These observations suggest that the Nested LSTM model, in particular, doesn’t use long-term memorization as one would otherwise expect from the additional memory states.

The observations in the article make no attempt to show the trend you're inferring from a single example. That's entirely fine for showcasing a visualization method, but it feels like too big a claim to generalize from this one example. Together with the fact that your nested LSTM model wasn't trained to comparable accuracy, I expect this to not sit well with Peer Reviewers and would propose a more cautious wording.

## Consider a rewritten subtitle

Take the current subtitle:

"When do Recurrent Neural Networks use short-term memorization or long-term memorization, and how do different architectures compare in terms of this? Visualizing the gradient can be a powerful tool to answer these questions."

- rather than comparing entire architectures the article specifically focusses on the recurrent units, which seems worth mentioning as it's a much more specific space to explore.
- Second sentence could be more specific about what "Visualizing the gradient" helps achieve.
- It's a little long.

A possible rewriting:

"When do recurrent neural networks use short-term or long-term memorization? By looking at gradient magnitudes in context, we can see differences in how recurrent units use memorization."

You could even consider leaving out the first sentence .

## Consider an even tighter introduction

The current introduction reads to me like it's trying to cover everything the article will talk about. This is an understandable angle, given that Introductions in Distill articles often feel like abstracts. They are not, however, and I'd encourage you to focus on the main contribution of the article: visualizing connectivity in recurrent architectures.

Let's be more concrete:

* Second paragraph: (Nested LSTM)
 It feels sudden and unexpected to mention Nested LSTM here. The brief survey in "Recurrent Units" seems like a better place for that.

* Third paragraph: (Inspiration)
 The "Comparing Recurrent Units" seems like the place to mention this as prior work.
 
* Fourth paragraph: (What this article is about)
 This feels closest to what we expect from an introduction. I'd still move the explanation of why the autocomplete problem is used to the "Comparing Recurrent Units" section.
 
* Fifth paragraph: (advance dialogue)
 This seems to serve the editors more than the readers, I'd recommend to drop entirely.
 
* What feels like its missing:
 A single sentence describing the two-section structure of the article.
 
 
## Consider annotated math in "Recurrent Units"

While the introductory formulas are relatively easy to understand and likely conceptually known to most readers, it seems like a really good opportunity to use annotated formulas. When most of the work a formula accomplishes is assigning names, it is particularly helpful to keep symbols spatially close to their definitions.

![image](https://cloud.githubusercontent.com/assets/61658/22921158/5ad92038-f24d-11e6-932f-fbc55cc49f2a.png)

View an [example in the momentum article](https://distill.pub/2017/momentum/#momentum_annotations). You can achieve this effect with CSS positioning, or by drawing the entire figure in SVG.

## Consider small changes to the "Recurrent Units" diagram

I like how this diagram gets across the main message of vanishing gradients really effectively! A tiny bit (maybe 2-3px?) of border-radius (or `rx` in SVG) could help to make the `rects` look a bit lighter. What if the arrows of future timesteps aren't blue yet, but grayed out?

![image](https://storage.googleapis.com/ludwigschubert/images-misc/vanishing-gradient-arrows-and-rounded-boxes.png)

I'd suggest saving the formula of the vanilla RNN gradient in the `figcaption` formulas for a footnote, the longer explanation of RNNs, or simply leaving them out.

## Consider rephrasing "Recurrent Units" introduction

This is a nitpick: "because of this time dependency it is very common for the network to suffer from a vanishing gradient problem" is one of those "weasel-word" constructions where even though we use "because" no actual explanation follows. I'd suggest rephrasing as either neutrally—"[time-dependency] causes [vanishing gradient problem]"—or actually talking about how at each time step the gradient is multiplied by a value smaller than one, and how that follows from the activation functions used.

## Consider moving accuracy table to a footnote

Our suggestion here would be to mention the lower accuracy of the Nested LSTM in order to hedge the article's claims, but not to show the table in the main article flow. (So either as a footnote or included in the already very nice appendix.)

## Parts we really enjoyed

Since receiving many items of feedback all at once can sometimes feel like there are no upsides to one's work, I wanted to include parts we really enjoyed seeing, too!

* Switchable x-axis from time to epochs. Now this is using web publishing to its fullest! :-)

* Explorable diagram: While gradient magnitudes could have been displayed in a single 2D diagram all at once, I liked the interactive version, too: it gets readers to think more about why the gradient magnitudes change in the way they do. (One could also have and compare both!)

* Subtle interactivity: the vanishing gradient diagram works without interactivity, but is subtly enhanced by it.

* Interactive diagrams come with hints of which parts to focus on; this ameliorates one of the biggest issues with explorable diagrams: readers can get lost in them and miss interesting points.
